Mathlib/RingTheory/Binomial.lean

```diff
@@ -0,0 +1,124 @@
+/-
+Copyright (c) 2023 Scott Carnahan. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Carnahan
+-/
+
+import Mathlib.Data.Rat.NNRat
+import Mathlib.RingTheory.Polynomial.Pochhammer
+
+/-!
+# Binomial semirings and binomial rings
+
+In this file we introduce binomial semirings and binomial rings, together with related functions
+like binomial coefficients.
+
+According to our main reference [Elliott] (which lists many equivalent conditions), a binomial ring
+is a torsion-free commutative ring R such that for any x ∈ R and any k ∈ ℕ, the product
+x(x-1)⋯(x-k+1) ∈ (k!)R.  The torsion-free condition lets us divide by k! unambiguously, so we get
+uniquely defined binomial coefficients.
+
+The defining condition doesn't require commutativity, and we get a theory with essentially the same
+power by replacing subtraction with addition.  Thus, we consider a semiring R in which
+multiplication by factorials is injective, and demand that for all x ∈ R and any k ∈ ℕ,
+k! | x(x+1)⋯(x+(k-1)).
+The quotient is called "multichoose r k", following the convention given for natural numbers.
+
+## Main results
+
+none yet
+
+## References
+
+J. Elliott, `Binomial rings, integer-valued polynomials, and λ-rings` J. Pure Appl. Alg. 207 (2006)
```

Comment:
We actually have a nice references system and a bibfile (e.g https://github.com/leanprover-community/mathlib4/blob/27866b848920f9620ed5a72c9039fff58e479d6f/Mathlib/NumberTheory/FunctionField.lean#L34) can you format this using that instead?

(Ok it doesn't actually work yet with the new doc-gen but it should at some point in the future I assume and at that point it will be good if this is formatted properly)

Reply:
Yes, I think the new revision is in the right format.